R/db-odbc-access.R

```diff
@@ -0,0 +1,240 @@
+                               x, "* 2 * ATN(1), ",                                          # Return pi/2 (if 1) or -pi/2 (if -1) using trig manipulation
+                               "ATN(", x, "/ SQR(-", x, "*", x, " + 1)))")                   # Otherwise return ASIN(x) using trig manipulation of ATAN
+                   },
+                   # Inverse hyperbolic sine for the domain x > 1 satisfies log(x + sqrt(x^2 + 1)). x <= 1 returns complex results
+                   asinh         = function(x) {
+                     build_sql("LOG(", x, " + SQR(", x, "^2 + 1))")
+                   },
+                   atan          = sql_prefix("ATN"),
+                   # Note that R takes y then x as arguments
+                   atan2         = function(y, x) {
+                     build_sql("IIF(", y, " > 0, ",
+                                    "IIF(", x, " >= ", y, ", ",
+                                         "ATN(", y, " / ", x, "), ",
+                                         "IIF(", x, " <= -", y, ", ",
+                                              "ATN(", y, " / ", x, "+ ATN(1) * 4), ",
+                                              "ATN(1) * 2 - ATN(", x, " / ", y, "))), ",
+                                    "IIF(", x, ">= -", y, ", ",
+                                         "ATN(", y, "/", x, "), ",
+                                         "IIF(", x, "<=", y, ", ",
+                                              "ATN(", y, "/", x, "- ATN(1)*4), ",
+                                              "-ATN(1)*2 - ATN(", x, "/", y, "))))")
+                   },
+                   # Inverse hyperbolic tan for the domain -1 < x < 1 satisfies log((1 + x) / (1 - x)) / 2.
+                   atanh         = function(x) {
+                     build_sql("LOG((1 + ", x, ") / (1 - ", x, ")) / 2")
+                   },
+                   cot           = function(x) {
+                     build_sql("1 / TAN(", x, ")")
+                   },
+                   coth          = function(x) {
+                     build_sql("(EXP(", x, ") + EXP(-", x, ")) / (EXP(", x, ") - EXP(-", x, "))")
+                   },
+
+
+                   # Strings
+                   nchar         = sql_prefix("LEN"),
+                   tolower       = sql_prefix("LCASE"),
+                   toupper       = sql_prefix("UCASE"),
+                   # Pull `left` chars from the left, then `right` chars from the right to replicate substr
+                   substr        = function(x, start, stop){
+                     right  <- stop - start + 1
+                     left   <- stop
+                     build_sql(
+                       "RIGHT(LEFT(", x, ", ", left, "), ", right, ")")
+                   },
+                   trimws        = sql_prefix("TRIM"),
+                   # No support for CONCAT in Access
+                   paste         = sql_not_supported("paste()"),
```

Comment:
This is either more challenging than I originally thought, or I am missing something easy. 

```r
sql_paste <- function(..., sep = " ", collapse = NULL) {
  binary_build <- function(x, y) {
    build_sql(x, " & ", sep, " & ", y)
  }

  .dots <- list(...)
  last_dot <- length(.dots)

  if(last_dot == 1) {
    build_sql("CStr(", .dots[[1]], ")")
  } else {
    reduce(.dots, binary_build)
  }

}

sql_paste0 <- function(..., collapse = NULL) {
  ex_paste(..., sep = "")
}
```

Something like this does work. The challenge is building correct SQL for an arbitrary number of inputs. This `reduce()` method works, but if you have other suggestions I'm open to them.

Reply:
I'd just do 

```
  if (!is.null(collapse)) {
    stop("`collapse` isn't supported in SQL translation", call. = FALSE)
  }
```

Since a non-NULL `collapse` turns it from a scalar function to an aggregate function.

Otherwise, the `reduce()` looks fine. I think `last_dot` would be better called `n`.

Can you please call the function something like `sql_paste_access()` and define outside of the list?

Reply:
Done!